consensus/src/liveness/leader_reputation.rs

```diff
@@ -202,18 +187,186 @@ impl ReputationHeuristic for ActiveInactiveHeuristic {
+pub struct ActiveInactiveHeuristic {
+    author: Author,
+    active_weight: u64,
+    inactive_weight: u64,
+    aggregation: NewBlockEventAggregation,
+}
+
+impl ActiveInactiveHeuristic {
+    pub fn new(
+        author: Author,
+        active_weight: u64,
+        inactive_weight: u64,
+        window_size: usize,
+    ) -> Self {
+        Self {
+            author,
+            active_weight,
+            inactive_weight,
+            aggregation: NewBlockEventAggregation::new(window_size, window_size),
+        }
+    }
+}
+
+impl ReputationHeuristic for ActiveInactiveHeuristic {
+    fn get_weights(
+        &self,
+        epoch: u64,
+        candidates: &[Author],
+        history: &[NewBlockEvent],
+    ) -> Vec<u64> {
+        let votes = self.aggregation.count_votes(epoch, candidates, history);
+        let proposals = self.aggregation.count_proposals(epoch, history);
+
+        COMMITTED_PROPOSALS_IN_WINDOW.set(*proposals.get(&self.author).unwrap_or(&0) as i64);
+        COMMITTED_VOTES_IN_WINDOW.set(*votes.get(&self.author).unwrap_or(&0) as i64);
+        LEADER_REPUTATION_WINDOW_SIZE.set(proposals.values().sum::<i32>() as i64);
+
+        candidates
+            .iter()
+            .map(|author| {
+                if votes.contains_key(author) || proposals.contains_key(author) {
+                    self.active_weight
+                } else {
+                    self.inactive_weight
+                }
+            })
+            .collect()
+    }
+}
+
+/// Heuristic that looks at successful and failed proposals, as well as voting history,
+/// to define node reputation, used for leader selection.
+///
+/// We want to optimize leader selection to primarily maximize network's throughput,
+/// but we also, in combinatoin with staking rewards logic, need to be reasonably fair.
+///
+/// Logic is:
+///  * if proposer round failure rate within the proposer window is strictly above threshold, use failed_weight (default 1).
+///  * otherwise, if node had no proposal rounds and no successful votes, use inactive_weight (default 10).
+///  * otherwise, use the default active_weight (default 100).
+///
+/// We primarily want to avoid failed rounds, as they have a largest negative effect on the network.
+/// So if we see a node having failures to propose, when it was the leader, we want to avoid that node.
+/// We add a threshold (instead of penalizing on a single failure), so that transient issues in the network,
+/// or malicious behaviour of the next leader is avoided. In general, we expect there to be
+/// proposer_window_size/num_validators opportunities for a node to be a leader, so a single failure, or a
+/// subset of following leaders being malicious will not be enough to exclude a node.
+/// On the other hand, single failure, without any successes before will exclude the note.
+/// Threshold probably makes the most sense to be between:
```

Comment:
One problem with the threshold is that malicious leaders who want to slowdown the network can craft proposal failures in a way that they can slowdown the network but still stay under the radar so that they are not penalized.  It actually depends on the threshold value itself but I am wondering why not use % of failures as a weight instead of a fixed threshold?